packages/ttypescript/src/PluginCreator.ts

```diff
@@ -34,7 +34,24 @@ export interface PluginConfig {
     afterDeclarations?: boolean;
 }
 
-export interface TransformerBasePlugin {
+export type CreateProgramMiddlewareNext = (createProgramOptions?: ts.CreateProgramOptions) => ts.Program;
+export type CreateProgramMiddlewareHead = (createProgramOptions: ts.CreateProgramOptions) => ts.Program;
+export type CreateProgramMiddleware = (createProgramOptions: ts.CreateProgramOptions, next: CreateProgramMiddlewareNext) => ts.Program;
+
+declare module 'typescript' {
+    export interface Middleware {
+        createProgram?: CreateProgramMiddleware;
```

Comment:
Why `createProgram` can be optional here?

Reply:
In future, if more methods to intercept are added, it might be undesirable to require a module to implement no-op forwarders for all methods that are there. Also requiring every method would break existing middleware modules, as their export signature would no longer match with what `ttypescript` expects.

I'd suppose that in future this interface would become something like:

```ts
export interface Middleware {
    createProgram?: CreateProgramMiddleware;
    resolveModuleName?: ResolveModuleNameMiddleware;
    transpileModule?: TranspileModuleMiddleware;
    // etc...
}
```

And implementors would be able to choose what functions are they interested in.